File: backend/api/views/geocoding.py
"""
Geocoding proxy view to handle reverse geocoding requests
Avoids CORS and 403 issues by proxying through backend
"""
import requests
import random
from rest_framework.decorators import api_view, permission_classes
from rest_framework.permissions import AllowAny
from rest_framework.response import Response
from rest_framework import status
from django.core.cache import cache
import time

# Pool of user agents
USER_AGENTS = [
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:121.0) Gecko/20100101 Firefox/121.0',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
]

def get_random_user_agent():
    """Get a random user agent from the pool"""
    return random.choice(USER_AGENTS)

def respect_rate_limit():
    """Ensure we don't exceed 1 request per second to Nominatim"""
    last_request_key = 'nominatim_last_request'
    last_request_time = cache.get(last_request_key, 0)
    current_time = time.time()
    
    time_since_last = current_time - last_request_time
    if time_since_last < 1.0:
        time.sleep(1.0 - time_since_last)
    
    cache.set(last_request_key, time.time(), timeout=10)

@api_view(['GET'])
@permission_classes([AllowAny])
def reverse_geocode(request):
    """
    Proxy endpoint for reverse geocoding
    GET /api/geocoding/reverse/?lat=<latitude>&lon=<longitude>
    """
    lat = request.GET.get('lat')
    lon = request.GET.get('lon')
    
    if not lat or not lon:
        return Response(
            {'error': 'Missing required parameters: lat and lon'},
            status=status.HTTP_400_BAD_REQUEST
        )
    
    try:
        # Validate coordinates
        lat_float = float(lat)
        lon_float = float(lon)
        
        if not (-90 <= lat_float <= 90) or not (-180 <= lon_float <= 180):
            return Response(
                {'error': 'Invalid coordinates'},
                status=status.HTTP_400_BAD_REQUEST
            )
    except ValueError:
        return Response(
            {'error': 'Invalid coordinate format'},
            status=status.HTTP_400_BAD_REQUEST
        )
    
    # Check cache first (cache for 1 hour)
    cache_key = f'geocode_{lat}_{lon}'
    cached_result = cache.get(cache_key)
    if cached_result:
        return Response(cached_result)
    
    # Try Nominatim first
    try:
        respect_rate_limit()
        
        nominatim_url = f'https://nominatim.openstreetmap.org/reverse'
        params = {
            'format': 'json',
            'lat': lat,
            'lon': lon,
            'zoom': 18,
            'addressdetails': 1
        }
        headers = {
            'User-Agent': get_random_user_agent(),
            'Accept': 'application/json',
            'Accept-Language': 'en',
            'Referer': 'https://smartwayz.app'
        }
        
        response = requests.get(nominatim_url, params=params, headers=headers, timeout=10)
        
        if response.status_code == 200:
            data = response.json()
            
            # Build detailed address from components
            if data.get('address'):
                addr = data['address']
                parts = [
                    addr.get('road') or addr.get('street'),
                    addr.get('suburb') or addr.get('neighbourhood'),
                    addr.get('city') or addr.get('town') or addr.get('municipality') or addr.get('village'),
                    addr.get('state') or addr.get('province'),
                    addr.get('country')
                ]
                address = ', '.join([p for p in parts if p])
            else:
                address = data.get('display_name', '')
            
            result = {
                'address': address,
                'provider': 'nominatim',
                'raw': data
            }
            
            # Cache the result
            cache.set(cache_key, result, timeout=3600)
            
            return Response(result)
    
    except Exception as e:
        print(f"Nominatim failed: {str(e)}")
    
    # Fallback to BigDataCloud
    try:
        bdc_url = f'https://api.bigdatacloud.net/data/reverse-geocode-client'
        params = {
            'latitude': lat,
            'longitude': lon,
            'localityLanguage': 'en'
        }
        
        response = requests.get(bdc_url, params=params, timeout=10)
        
        if response.status_code == 200:
            data = response.json()
            
            # Build address from components
            parts = []
            if data.get('localityInfo', {}).get('administrative'):
                admin = data['localityInfo']['administrative']
                if len(admin) > 6:
                    parts.append(admin[6].get('name'))
                if len(admin) > 5:
                    parts.append(admin[5].get('name'))
            
            parts.extend([
                data.get('locality') or data.get('city'),
                data.get('principalSubdivision'),
                data.get('countryName')
            ])
            
            address = ', '.join([p for p in parts if p])
            
            result = {
                'address': address,
                'provider': 'bigdatacloud',
                'raw': data
            }
            
            # Cache the result
            cache.set(cache_key, result, timeout=3600)
            
            return Response(result)
    
    except Exception as e:
        print(f"BigDataCloud failed: {str(e)}")
    
    # Last resort: return coordinates
    result = {
        'address': f'Location: {lat}, {lon}',
        'provider': 'coordinates',
        'raw': None
    }
    
    return Response(result)

File: frontend/src/services/geocoding.js
// Backend API URL
const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000/api';

/**
 * Try backend proxy for geocoding (handles Nominatim and fallbacks)
 */
const tryBackendProxy = async (lat, lng) => {
  const response = await fetch(
    `${API_BASE_URL}/geocoding/reverse/?lat=${lat}&lon=${lng}`
  );
  
  if (!response.ok) {
    throw new Error(`Backend proxy HTTP ${response.status}`);
  }
  
  const data = await response.json();
  
  if (data.address) {
    return data.address;
  }
  
  throw new Error('No address found');
};

/**
 * Try OpenCage Geocoding API (free tier: 2500 requests/day)
 * Note: Requires API key - user needs to sign up at opencagedata.com
 */
const tryOpenCage = async (lat, lng) => {
  const apiKey = import.meta.env.VITE_OPENCAGE_API_KEY;
  
  if (!apiKey) {
    throw new Error('OpenCage API key not configured');
  }
  
  const response = await fetch(
    `https://api.opencagedata.com/geocode/v1/json?q=${lat}+${lng}&key=${apiKey}`
  );
  
  if (!response.ok) {
    throw new Error(`OpenCage HTTP ${response.status}`);
  }
  
  const data = await response.json();
  
  if (data.results && data.results.length > 0) {
    return data.results[0].formatted;
  }
  
  throw new Error('No address found');
};

/**
 * Try BigDataCloud reverse geocoding (free, no API key required)
 */
const tryBigDataCloud = async (lat, lng) => {
  const response = await fetch(
    `https://api.bigdatacloud.net/data/reverse-geocode-client?latitude=${lat}&longitude=${lng}&localityLanguage=en`
  );
  
  if (!response.ok) {
    throw new Error(`BigDataCloud HTTP ${response.status}`);
  }
  
  const data = await response.json();
  
  // Build detailed address from all available components
  const parts = [
    data.localityInfo?.administrative?.[6]?.name, // Street/Road
    data.localityInfo?.administrative?.[5]?.name, // Barangay/District
    data.locality || data.city,                    // City/Municipality
    data.principalSubdivision,                     // Province/State
    data.countryName                               // Country
  ].filter(Boolean);
  
  if (parts.length > 0) {
    return parts.join(', ');
  }
  
  throw new Error('No address found');
};

/**
 * Fallback: Generate simple coordinate-based address
 */
const generateCoordinateAddress = (lat, lng) => {
  return `Location: ${lat.toFixed(6)}, ${lng.toFixed(6)}`;
};

/**
 * Main geocoding function with multiple fallbacks
 * @param {number} lat - Latitude
 * @param {number} lng - Longitude
 * @returns {Promise<string>} Address string
 */
export const reverseGeocode = async (lat, lng) => {
  const providers = [
    { name: 'Backend Proxy', fn: tryBackendProxy },
    { name: 'BigDataCloud', fn: tryBigDataCloud },
    { name: 'OpenCage', fn: tryOpenCage }
  ];
  
  // Try each provider in order
  for (const provider of providers) {
    try {
      console.log(`Trying ${provider.name} geocoding...`);
      const address = await provider.fn(lat, lng);
      console.log(`✓ ${provider.name} succeeded:`, address);
      return address;
    } catch (error) {
      console.warn(`✗ ${provider.name} failed:`, error.message);
      // Continue to next provider
    }
  }
  
  // All providers failed, use coordinate fallback
  console.log('All geocoding providers failed, using coordinates');
  return generateCoordinateAddress(lat, lng);
};

export default reverseGeocode;
